dm/dm/master/scheduler/scheduler.go

```diff
@@ -2362,51 +2279,54 @@ func (s *Scheduler) tryBoundForSource(source string) (bool, error) {
 				continue
 			}
 			// the worker is not Offline
-			if w.Stage() == WorkerRelay {
+			if w.Stage() != WorkerOffline && len(w.Bounds()) < boundSourcesNum {
 				worker = w
+				boundSourcesNum = len(w.Bounds())
 				s.logger.Info("found relay worker when source bound",
 					zap.String("worker", workerName),
 					zap.String("source", source))
-				break
 			}
 		}
 	}
 	// then a history worker for this source...
+	boundSourcesNum = sourceNum
 	if worker == nil {
 		for _, bound := range s.lastBound {
-			workerName := bound.Worker
 			if bound.Source == source {
-				w, ok := s.workers[workerName]
+				w, ok := s.workers[bound.Worker]
 				if !ok {
-					// a not found worker
-					continue
+					// a not found worker, should not happen
+					s.logger.DPanic("worker instance not found for history worker", zap.String("worker", bound.Worker))
+					return false, nil
 				}
-				if w.Stage() == WorkerFree {
+
+				if w.Stage() != WorkerOffline && len(w.Bounds()) < boundSourcesNum {
 					worker = w
+					boundSourcesNum = len(w.Bounds())
 					s.logger.Info("found history worker when source bound",
 						zap.String("worker", workerName),
 						zap.String("source", source))
-					break
 				}
 			}
 		}
 	}
 
-	// and then a random Free worker.
+	// and then a Free worker with lease sources.
+	boundSourcesNum = sourceNum
 	if worker == nil {
 		for _, w := range s.workers {
-			if w.Stage() == WorkerFree {
+			if w.Stage() != WorkerOffline && len(w.Bounds()) < boundSourcesNum {
```

Comment:
What about taking other decisive factors into account in addition to `boundSourcesNum`? Considering some workers might be bounded to a lot of sources but have few tasks to run, will it be better to bound more sources to these "free" workers? (or maybe leave a todo here)🤔

Reply:
In this phase, we only decide this through sources number. I will add a TODO here. Maybe we will use task numbers or source QPS load in the future.